src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+                                        qop + Colon +
+                                        ComputeHash(a2, algorithm), algorithm);
+
+            // Add response
+            sb.AppendKeyValue(Response, response);
+
+            // Add algorithm
+            sb.AppendKeyValue(Algorithm, algorithm, includeQuotes: false);
+
+            // Add opaque
+            sb.AppendKeyValue(Opaque, digestResponse.Parameters[Opaque]);
+
+            // Add qop
+            sb.AppendKeyValue(Qop, qop, includeQuotes: false);
+
+            // Add nc
+            sb.AppendKeyValue(NC, digestResponse.NonceCount.ToString("x8"), includeQuotes: false);
+
+            // Add cnonce
+            sb.AppendKeyValue(CNonce, cnonce, includeComma: false);
+
+            return StringBuilderCache.GetStringAndRelease(sb);
+        }
+
+        private static string GetRandomAlphaNumericString()
+        {
+            using (RNGCryptoServiceProvider rng = new RNGCryptoServiceProvider())
+            {
+                const int length = 16;
+                StringBuilder sb = StringBuilderCache.Acquire(length);
+                for (int i = 0; i < length; i++)
+                {
+                    byte[] randomNumber = new byte[1];
+                    rng.GetBytes(randomNumber);
+
+                    int rangeIndex = (randomNumber[0] % 3);
+                    rng.GetBytes(randomNumber);
+
+                    if (rangeIndex == 0)
+                    {
+                        // Get a random digit 0-9
+                        sb.Append((char)(AlphaNumChooser[rangeIndex] + randomNumber[0] % 10));
+                    }
+                    else
+                    {
+                        // Get a random alphabet in a-z, A-Z
+                        sb.Append((char)(AlphaNumChooser[rangeIndex] + randomNumber[0] % 26));
+                    }
+                }
+
+                return StringBuilderCache.GetStringAndRelease(sb);
+            }
+        }
+
+        private static string ComputeHash(string data, string algorithm)
+        {
+            if (algorithm == Sha256)
+            {
+                using (HashAlgorithm hash = SHA256.Create())
+                {
+                    return ComputeHash(data, hash);
+                }
+            }
+            else
+            {
+// Disable MD5 insecure warning.
+#pragma warning disable CA5351
+                using (HashAlgorithm hash = MD5.Create())
+                {
+                    return ComputeHash(data, hash);
+                }
+#pragma warning restore CA5351
+            }
+        }
+
+        private static string ComputeHash(string data, HashAlgorithm hash)
+        {
+            Encoding enc = Encoding.UTF8;
+            byte[] result = hash.ComputeHash(enc.GetBytes(data));
+
+            StringBuilder sb = StringBuilderCache.Acquire(result.Length * 2);
+            foreach (byte b in result)
+                sb.Append(b.ToString("x2"));
```

Comment:
Lots of places we'll be able to reduce allocations once we have the span support in place.